security/fma-l1-upgrades.md

```diff
@@ -0,0 +1,306 @@
+
+This document covers the L1 Upgrades project, which aims to improve and standardize the approach to upgrading L1 contracts across the superchain. The project introduces a new upgrade system that consolidates ProxyAdmin contracts, implements a two-step upgrade flow, and provides a more deterministic and safer upgrade process.
+
+Below are references for this project:
+
+- [L1 Upgrades Design Doc](../protocol/l1-upgrades.md)
+- [OPCM spec](https://github.com/ethereum-optimism/specs/blob/opcm/upgrades/specs/experimental/op-contracts-manager.md)
+
+### FM1: Contract Initializers
+
+#### **Description:**
+
+During the first step of the ugprade, the contracts will not yet have a shared
+storage location for the `initialized` value. Therefore each contract will require custom storage
+manipulation to properly reset the `initialized` value, without modifying values which may be
+packed into the same slot.
+
+Moreoever, contracts could be re-initialized if the initialization state is not properly managed
+during the two-step upgrade process.
+
+#### **Risk Assessment:**
+
+High impact, Medium likelihood
+
+#### **Mitigations:**
+
+1. In the first step to reset the initializer with varied layouts, a custom contract can be
+   created which hardcodes the exact modifications required for each contract with a unique
+   contracts specific function, such as `resetL1CrossDomainMessenger()`.
+2. Upgrading to use OpenZeppelin's unstructured storage for initialization state reduces the risk
+   of incorrectly reinitializing at the end of the upgrade, as it ensures that the `initializer`
+   modifier can operate on the same slot across contracts.
+3. After calling `OPCM.upgrade()` the `superchain-ops` script can use `vm.load()` and attempt to call `L1CrossDomainMessenger.upgrade()` (for example) to verify the intitialized status.
+4. Tenderly simulation should look for unexpected storage changes.
+
+#### **Detection:**
+
+Global event monitoring COULD be used on `monitorism` for the `Initialized` event, howevrer in the past similar monitoring
+has been deemed to noisy and not useful.
+
+#### **Recovery Path(s)**:
+
+An emergency upgrade would be required.
+
+### FM2: Storage Layout Corruption in an OP Chain Contract
+
+#### **Description:**
+
+During an upgrade, incorrect storage layout modifications could lead to
+corruption of contract state. This could happen if the upgrade process doesn't
+properly account for existing storage slots or if there are conflicts in the
+storage layout between versions.
+
+We disregard issues with the initializer as they are covered in FM1.
+
+#### **Risk Assessment:**
+
+High impact, Low likelihood
+
+#### **Mitigations:**
+
+Existing mitigations for this failure mode include:
+
+1. The use of spacer variables in L1 contracts to prevent storage collisions.
+2. The use of storage layout lock files tracked in git to detect breaking changes.
+3. The use of off-chain checks in the `superchain-ops` scripts to ensure that getters
```

Comment:
> Are you talking about a check that looks something like [this](https://github.com/ethereum-optimism/superchain-ops/blob/main/tasks/eth/020-set-gas-target/SignFromJson.s.sol?rgh-link-date=2024-12-04T16%3A10%3A21Z#L36)?

Yes.

> Am I right in saying we will continue to use the same superchain-ops process once we upgrade to opcm?

I think this is an open question, and one that the upgrades team will need to make together. I know there are significant changes to the ops repo planned, so that will have an impact here.